safe_url_check/lib/src/safe_url_check.dart

```diff
@@ -0,0 +1,144 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'dart:io';
+
+import 'package:retry/retry.dart';
+
+import 'private_ip.dart';
+import 'unique_local_ip.dart';
+import 'version.dart';
+
+const defaultUserAgent = 'package:safe_url_check/$packageVersion '
+    '(+https://github.com/google/dart-neats/tree/master/safe_url_check)';
+
+Future<bool> doSafeUrlCheck(
+  Uri url, {
+  int maxRedirects = 8,
+  String userAgent = defaultUserAgent,
+  HttpClient? client,
+  RetryOptions retryOptions = const RetryOptions(maxAttempts: 3),
+  Duration timeout = const Duration(seconds: 90),
+}) async {
+  ArgumentError.checkNotNull(url, 'url');
+  ArgumentError.checkNotNull(maxRedirects, 'maxRedirects');
+  ArgumentError.checkNotNull(userAgent, 'userAgent');
+  ArgumentError.checkNotNull(retryOptions, 'retryOptions');
+  if (maxRedirects < 0) {
+    throw ArgumentError.value(
+      maxRedirects,
+      'maxRedirects',
+      'must be a positive integer',
+    );
+  }
+
+  try {
+    // Create client if one wasn't given.
+    var c = client;
+    c ??= HttpClient();
+    try {
+      return await SafeUrlChecker(
+        client: c,
+        userAgent: userAgent,
+        retryOptions: retryOptions,
+        timeout: timeout,
+      ).checkUrl(url, maxRedirects);
+    } finally {
+      // Close client, if it was created here.
+      if (client == null) {
+        c.close(force: true);
+      }
+    }
+  } on Exception {
+    return false;
+  }
+}
+
+class SafeUrlChecker {
+  final HttpClient client;
+  final String userAgent;
+  final RetryOptions retryOptions;
+  final Duration timeout;
+  final bool skipLocalNetworkCheck;
+
+  SafeUrlChecker({
+    required this.client,
+    required this.userAgent,
+    required this.retryOptions,
+    required this.timeout,
+    this.skipLocalNetworkCheck = false,
```

Comment:
```suggestion
    @visibleForTestingOnly
    this.skipLocalNetworkCheck = false,
```

Reply:
This will pull in `package:meta` as a dependency. I don't mind, just wanted to note.